### PR TITLE
refactor(core): Improve test-webhooks (no-changelog)

### PR DESCRIPTION
**Pull request by ivov**
Remove duplication, improve readability, and expand tests for `TestWebhooks.ts` - in anticipation for storing test webhooks in Redis.

### CODE REVIEW

File: packages/cli/src/ActiveWebhooks.ts [hunk 1]
```diff
@@ -6,8 +6,9 @@ import type {
 	WorkflowActivateMode,
 	WorkflowExecuteMode,
 } from 'n8n-workflow';
-import { ApplicationError, WebhookPathTakenError } from 'n8n-workflow';
+import { WebhookPathTakenError } from 'n8n-workflow';
 import * as NodeExecuteFunctions from 'n8n-core';
+import { IdLessWorkflowError } from './errors/id-less-workflow.error';
 
 @Service()
 export class ActiveWebhooks {
```
Comment (netroy) [inline packages/cli/src/ActiveWebhooks.ts:14]: while we are at it, can you please move all of `ActiveWebhooks` into `TestWebhooks` and delete this file?
I'd like to create another `ActiveWebhooks` at some point to break out al webhooks code from `ActiveWorkflowRunner`, but that can wait for now.

**Review by netroy — CHANGES_REQUESTED** [review]

File: packages/cli/src/errors/id-less-workflow.error.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import type { Workflow } from 'n8n-workflow';
+import { ApplicationError } from 'n8n-workflow';
+
+export class IdLessWorkflowError extends ApplicationError {
```
Comment (netroy) [inline packages/cli/src/errors/id-less-workflow.error.ts:4]: Would `WorkflowMissingIdError` be a better name for this?

File: packages/cli/src/TestWebhooks.ts [hunk 3]
```diff
@@ -22,30 +21,40 @@ import type {
...
+	private testWebhooks: {
 		[key: string]: {
 			sessionId?: string;
 			timeout: NodeJS.Timeout;
-			workflowData: IWorkflowDb;
+			workflowEntity: IWorkflowDb;
 			workflow: Workflow;
 			destinationNode?: string;
 		};
 	} = {};
```
Comment (netroy) [inline packages/cli/src/TestWebhooks.ts:47]: Can we please extract the value type into a named interface, and change this to 
```
private registeredWebhooks: Record<string, RegisteredWebhook> = {}`
```

File: packages/cli/src/TestWebhooks.ts [hunk 4]
```diff
@@ -58,69 +67,54 @@ export class TestWebhooks implements IWebhookManager {
 		response: express.Response,
 	): Promise<IResponseCallbackData> {
 		const httpMethod = request.method;
-		let path = request.params.path;
 
-		// Reset request parameters
+		let path = request.params.path.endsWith('/')
+			? request.params.path.slice(0, -1)
+			: request.params.path;
+
```
Comment (netroy) [inline packages/cli/src/TestWebhooks.ts:74]: considering that we do this for test and prod webhooks, maybe we should move this bit to `WebhookHelpers.webhookRequestHandler`, and start passing a context object into `executeWebhook` calls.

**Review by netroy — CHANGES_REQUESTED** [review]
Thanks for doing this.
I'm seeing a bit of overlap between this and https://github.com/n8n-io/n8n/pull/7223, would like to cherry-pick some bits out of that PR in here if possible.

File: packages/cli/src/TestWebhooks.ts (hunk 4 above)
Comment (ivov) [reply packages/cli/src/TestWebhooks.ts:58]: Not quite familiar yet with `WebhookHelpers` to address quickly, I'll come back to this later if needed.

**Comment by ivov** [comment]
> Thanks for doing this. I'm seeing a bit of overlap between this and #7223, would like to cherry-pick some bits out of that PR in here if possible.

Please go ahead!

File: packages/workflow/src/Workflow.ts [hunk 5]
```diff
@@ -1009,7 +1009,7 @@ export class Workflow {
 		nodeExecuteFunctions: INodeExecuteFunctions,
 		mode: WorkflowExecuteMode,
 		activation: WorkflowActivateMode,
-		isTest?: boolean,
+		{ isTest }: { isTest?: boolean } = {},
```
Comment (netroy) [inline packages/workflow/src/Workflow.ts:1012]: we should either move more arguments into this object, or we should revert this change. having this just for `isTest` seems unnecessary.
or, maybe we could move `isTest` into `IWebhookData` ?

**Review by netroy — COMMENTED** [review]
besides the one comment, LGTM.

File: packages/workflow/src/Workflow.ts (hunk 5 above)
Comment (ivov) [reply packages/workflow/src/Workflow.ts:1012]: Thanks! I did this to avoid passing standalone booleans into the call - this arg used to be more readable as `this.testWebhooks`, until we merged `ActiveWebhooks.ts` into this file as part of the review.

I find `{ isTest: true }` more readable that keeping a standalone boolean. While I agree this legacy code needs much more work and we should consolidate options into a single arg, I wonder if that should that be done later down the line? The goal of this PR is prepping test webhooks so I can make the changes for the multi-main setup we discussed on Friday.

Comment (netroy) [reply packages/workflow/src/Workflow.ts:1012]: I generally agree, but the `isTest` changes in this PR seem like they are adding tech-debt. How about [something like this](https://github.com/n8n-io/n8n/commit/60c8a7f18e13797e5b34a4347fa8afe6ae321066)?

Comment (ivov) [reply packages/workflow/src/Workflow.ts:1012]: Ah I see what you mean now! If you can please bring this in.

**Review by netroy — APPROVED** [review]
👍🏽

**Comment by janober** [comment]
Got released with [`n8n@1.22.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.22.0)
